proposals/lip-0052.md

```diff
@@ -1520,14 +1520,15 @@ This function returns the next available index of a collection.
 
 ```python
 def getNextAvailableIndex(collectionID: CollectionID) -> uint64:
-
-    count = 0
-
-    for key in NFTStore.keys():
-        if key[`LENGTH_CHAIN_ID`:(`LENGTH_CHAIN_ID`+`LENGTH_COLLECTION_ID`)] == collectionID:
-            count +=1
-
-    return count
+    collectionNftIDs = sorted([key for key in NFTStore.keys() if key[:LENGTH_CHAIN_ID] == OWN_CHAIN_ID and key[LENGTH_CHAIN_ID:LENGTH_CHAIN_ID+LENGTH_COLLECTION_ID] == collectionID], reverse = True)
+    if len(collectionNftIDs) == 0:
+        return 0 
+    largestKey = collectionNftIDs[0]
+    largestIndex = int.from_bytes(largestKey[(LENGTH_CHAIN_ID + LENGTH_COLLECTION_ID):], byteorder='big')
+    if largestIndex == 2**64 - 1:
+        raise Exception("No more available indexes")
+    nextAvailableIndex = largestIndex + 1
+    return nextAvailableIndex
```

Comment:
For instance, here is a function in the [BFT implementation](https://github.com/LiskHQ/lisk-sdk/blob/98c3d9c588b56845e46eb8a209985a92bf478c67/framework/src/engine/bft/bft_params.ts#L21) which does something similar:
```
export const getBFTParameters = async (
	paramsStore: StateStore,
	height: number,
): Promise<BFTParameters> => {
	const start = utils.intToBuffer(0, 4);
	const end = utils.intToBuffer(height, 4);
	const results = await paramsStore.iterate({
		limit: 1,
		gte: start,
		lte: end,
		reverse: true,
	});
	if (results.length !== 1) {
		throw new BFTParameterNotFoundError();
	}
	const [result] = results;
	return codec.decode<BFTParameters>(bftParametersSchema, result.value);
};
```

Reply:
Thanks for the suggestion, I wan't sure how to do it. I will change the code accordingly

Reply:
This would be the high-level version:
```python
def getNextAvailableIndex(collectionID: CollectionID) -> uint64:
    start = chainID + collectionID + uint64be(0)
    end = chainID + collectionID + uint64be(2**64 - 1) 
    largestKey = lexicographically largest element in NFTStore with key between start and end
    if largestKey does not exist:
        return 0
    if largestIndex == 2**64 - 1:
        raise Exception("No more available indexes")
    nextAvailableIndex = largestIndex + 1
    return nextAvailableIndex
```

Reply:
This would be the high-level version:
```python
def getNextAvailableIndex(collectionID: CollectionID) -> uint64:
    start = chainID + collectionID + uint64be(0)
    end = chainID + collectionID + uint64be(2**64 - 1) 
    keys = NFTStore.iterate(
        limit: 1,
        gte: start,
        lte: end,
        reverse: true
    ) with key between start and end
    if length(keys) == 0:
        return 0
   largestIndex = keys[0]
    if largestIndex == 2**64 - 1:
        raise Exception("No more available indexes")
    return largestIndex + 1
```